### PR TITLE
New python lab

### PR DESCRIPTION
**Pull request by tussedrotten**

### CODE REVIEW

File: solution-segmentation.py [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+from sklearn.mixture import GaussianMixture
+
+
+class Rectangle:
+    """Represents a geometric rectangle"""
+
+    def __init__(self, top_left, bottom_right):
+        """Constructs a rectangle.
+
+        :param top_left: A tuple representing the top left point in the rectangle.
```
Comment (solosuper) [inline solution-segmentation.py]: Kan man være tydeligere på om det er (x,y) eller (y,x) (dvs row,col eller col,row ?

File: solution-segmentation.py [hunk 2]
```diff
@@ -0,0 +1,329 @@
+import cv2
+import numpy as np
+import scipy.spatial.distance as ssd
+from sklearn.mixture import GaussianMixture
+
+
+class Rectangle:
```
Comment (solosuper) [inline solution-segmentation.py]: Smart

File: solution-segmentation.py [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+    :param feature_image: An image of feature vectors.
+    :param sampling_rectangle: The region in the feature image to extract samples from.
+
+    :return The samples
+    """
+
+    patch = feature_image[sampling_rectangle.y_slice(), sampling_rectangle.x_slice()]
+    samples = patch.reshape(-1, 3)
+    return samples
```
Comment (solosuper) [inline solution-segmentation.py]: Spør for å lære: Er det riktig at man forstatt jobber på et "view" her, altså ikke tar noen kopi?

File: solution-segmentation.py [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    :param thresh: Threshold value.
+    :param use_otsu: Set to True to use Otsu's method to estimate the threshold value.
+    :param max_dist_value: The maximum distance value to represent after rescaling.
+
+    :return The updated threshold value and segmented image
+    """
+    # We need to represent the distances in uint16 because of OpenCV's implementation of Otsu's method.
+    scale = np.iinfo(np.uint16).max / max_dist_value
+    distances_scaled = np.uint16(np.clip(distance_image * scale, 0, np.iinfo(np.uint16).max))
+    thresh_scaled = thresh * scale
```
Comment (solosuper) [inline solution-segmentation.py]: Skulle denne også vært `np.clip`et?

File: solution-segmentation.py [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+
+
+def draw_sampling_rectangle(image, sampling_rectangle):
+    """Draw the sampling rectangle in an image
+
+    :param image: The image to draw the rectangle in.
+    :param sampling_rectangle: The sampling rectangle.
+    """
+
+    colour = (0, 0, 255)
```
Comment (solosuper) [inline solution-segmentation.py]: :gb:

File: solution-segmentation.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+    def compute_mahalanobis_distances(self, feature_image):
+        """Computes the Mahalanobis distances for a feature image given this model"""
+        samples = feature_image.reshape(-1, 3)
+        mahalanobis = ssd.cdist(samples, self.mean[None, :], metric='mahalanobis', VI=self.inverse_covariance)
+
+        return mahalanobis.reshape(feature_image.shape[:2])
+
+
+class GaussianMixtureModel:
```
Comment (solosuper) [inline solution-segmentation.py]: Kult

File: solution-segmentation.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+    # Connect to the camera.
+    # Change to video file if you want to use that instead.
+    device_id = 0
+    cap = cv2.VideoCapture(device_id)
+    if not cap.isOpened():
+        print(f"Could not open camera {device_id}")
+        return
+    else:
+        print(f"Successfully opened camera {device_id}")
```
Comment (solosuper) [inline solution-segmentation.py]: Siden du hinter til at man kan bruke videofil, hva med
```suggestion
    video_source = 0
    cap = cv2.VideoCapture(video_source)
    if not cap.isOpened():
        print(f"Could not open video source {video_source}")
        return
    else:
        print(f"Successfully opened video source {video_source}")
```

**Review by solosuper — APPROVED** [review]
Ser bra ut dette! Fornøyd med at jeg fortsatt kunne kjenne igjen endel av min kode, da var jeg kanskje ikke heelt på jordet :P

File: solution-segmentation.py (hunk 1 above)
Comment (tussedrotten) [reply solution-segmentation.py]: Kanskje du kan? :thinking:

File: solution-segmentation.py (hunk 3 above)
Comment (tussedrotten) [reply solution-segmentation.py]: Ja, jeg tror både `patch` og `samples` skal være "views" her

File: solution-segmentation.py (hunk 4 above)
Comment (tussedrotten) [reply solution-segmentation.py]: `thresh_scaled` er vel float her, så det går vel bra

File: lab_segmentation.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
+# Import libraries
+import cv2
+import numpy
+import numpy as np
+
+# Import common lab functions.
+from common_lab_utils import SegmentationLabGui, \
+    get_sampling_rectangle, draw_sampling_rectangle, extract_training_samples
```
Comment (solosuper) [inline lab_segmentation.py:7]: Dette ble jo veldig ryddig da. Greit å luke vekk alt som ikke har med segmentering å gjøre :+1:

File: common_lab_utils.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        # Create windows.
+        self.segm_win = 'Segmented frame'
+        self.maha_win = 'Mahalanobis image'
+        cv2.namedWindow(self.segm_win, cv2.WINDOW_NORMAL)
+        cv2.namedWindow(self.maha_win, cv2.WINDOW_NORMAL)
+
+        # Create slider that adjusts the threshold.
+        self.slider = 'Threshold'
+        thresh_setter_ref = self.__class__.thresh_val.__set__
+        cv2.createTrackbar(self.slider, self.segm_win, 0, max_thresh_val, lambda val: thresh_setter_ref(self, val))
```
Comment (solosuper) [inline common_lab_utils.py:23]: Watch and learn :clap:

File: common_lab_utils.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+        """Destroys the GUI"""
+        cv2.destroyWindow(self.segm_win)
+        cv2.destroyWindow(self.maha_win)
+
+    @property
+    def thresh_val(self):
+        """The threshold value"""
+        return self._thresh_val
+
+    @thresh_val.setter
```
Comment (solosuper) [inline common_lab_utils.py:42]: Artig å lære om properties

File: common_lab_utils.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+        # Create slider that adjusts the threshold.
+        self.slider = 'Threshold'
+        thresh_setter_ref = self.__class__.thresh_val.__set__
+        cv2.createTrackbar(self.slider, self.segm_win, 0, max_thresh_val, lambda val: thresh_setter_ref(self, val))
+
+        # Set threshold (and thereby trackbar) to initial value.
+        self._thresh_val = initial_thresh_val
+
+    def __enter__(self):
```
Comment (solosuper) [inline common_lab_utils.py:28]: Kult, lært enda en ny ting i dag :smiley: `with`

File: lab-guide/2-implement-simple-color-based-segmentation.md [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+- Hint: For a very efficient solution, take a look at [scipy.spatial.distance.cdist](https://docs.scipy.org/doc/scipy-1.8.0/html-scipyorg/reference/generated/scipy.spatial.distance.cdist.html?highlight=cdist#scipy-spatial-distance-cdist) 
+
+## 3. Experiment!
+Now you should have a working segmentation method, and it is finally time to play around with it!
+
+For example:
+- Try it out on different colours/surfaces. How well does it work?
+- Try changing the threshold manually using the slider.
+- Also check out how well Otsu's method estimates a decent threshold by pressing `o`.
```
Comment (solosuper) [inline lab-guide/2-implement-simple-color-based-segmentation.md]: Otsu krever jo en ganske bimodal fordeling, så hvis det er for få eller for mange forgrunnspiksler fungerer den ikke så bra. Er det noe man kan få dem til å tenke på? Når funker Otsu bra og når funker det dårlig?

File: lab_segmentation.py [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+                model = model_type(samples)
+
+            # Compute how well the pixel features fit with the model.
+            mahalanobis_img = model.compute_mahalanobis_distances(feature_image)
+
+            # Segment out the areas of the image that fits well enough.
+            gui.thresh_val, segmented = perform_segmentation(mahalanobis_img, gui.thresh_val, use_otsu, max_distance)
+
+            # Insert the segmented area as a green layer in the input frame and draw the sampling rectangle.
+            frame[segmented > 0] = (0, 255, 0)
```
Comment (solosuper) [inline lab_segmentation.py]: Hva med å sette rød og blå til 0 i stedet, og beholde intensitet i grønn? Jeg kan se på det hvis jeg får tid, og om det i tilfelle ser bra ut.

File: lab_segmentation.py [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+            draw_sampling_rectangle(frame, sampling_rectangle)
+
+            # Normalise the Mahalanobis image to [0, max_distance] for visualisation.
+            mahalanobis_img = mahalanobis_img / max_distance
```
Comment (solosuper) [inline lab_segmentation.py:70]: Kunne vi heller gjort det i `gui.show_mahalanobis`?

**Review by solosuper — COMMENTED** [review]
Jeg syns dette ser bra ut! Har ikke prøvd å kjøre koden hos meg.

File: lab-guide/2-implement-simple-color-based-segmentation.md (hunk 12 above)
Comment (tussedrotten) [reply lab-guide/2-implement-simple-color-based-segmentation.md]: Godt poeng, skriver det inn!

File: lab_segmentation.py (hunk 13 above)
Comment (tussedrotten) [reply lab_segmentation.py]: Ser kult ut, gjør det!

File: lab_segmentation.py (hunk 14 above)
Comment (tussedrotten) [reply lab_segmentation.py:70]: Beholder det, slik at studentene kan se alle bildeoperasjonene

**Review by solosuper — COMMENTED** [review]
Ser bra ut! Python solution kjører fint hos meg.

File: lab_segmentation.py [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+            mahalanobis_img = model.compute_mahalanobis_distances(feature_image)
+
+            # Segment out the areas of the image that fits well enough.
+            gui.thresh_val, segmented = perform_segmentation(mahalanobis_img, gui.thresh_val, use_otsu, max_distance)
+
+            # Highlight the segmented area in green in the input frame and draw the sampling rectangle.
+            frame[segmented > 0] *= np.uint8([0, 1, 0])
+            draw_sampling_rectangle(frame, sampling_rectangle)
+
+            # Normalise the Mahalanobis image to [0, max_distance] for visualisation.
```
Comment (thomasolsvikopsahl) [inline lab_segmentation.py]: Når du deler på max_distance så normaliserer du vel til [0,1] og ikke [0, max_distance]

File: lab_segmentation.py [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+
+        :param samples: A set of feature samples
+        """
+
+        self._perform_training(samples)
+
+    def _perform_training(self, samples):
+        """Trains the model"""
+
+        # TODO 1.1: Train the multivariate normal model by estimating the mean and covariance given the samples.
```
Comment (thomasolsvikopsahl) [inline lab_segmentation.py:115]: Bedre med "... for the given samples."?

File: lab_segmentation.py [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+        self._mean = np.ones(samples.shape[1])                       # Dummy solution, replace
+        self._covariance = np.identity(samples.shape[1])          # Dummy solution, replace
+
+        # TODO 1.2: Compute the inverse of the estimated covariance.
+        self._inverse_covariance = np.identity(samples.shape[1])  # Dummy solution, replace
```
Comment (thomasolsvikopsahl) [inline lab_segmentation.py]: Kjenner det rykke i OCD nerven når jeg ser at kommentarene nesten er alignet...

**Review by thomasolsvikopsahl — APPROVED** [review]
Bortsett fra noe uskyldig småplukk, så ser alt greit ut for meg.

File: lab_segmentation.py (hunk 16 above)
Comment (tussedrotten) [reply lab_segmentation.py:115]: Nope ;)
